### ISSUE TITLE
Nonintuitive behavior around commands accepting Iterables

### ISSUE THREAD

**Issue by coandco** [alisaifee/coredis#95]
I've recently been converting a small program over from using the old aredis library to using the latest coredis, and I hit what seemed at the time to be a pretty baffling error: when I tried to `client.lpush(keyname, string_to_push)`, coredis would push each letter in the string individually.  After some amount of tracking it down, I realized it was one of the breaking changes outlined in the [3.0.0 release notes](https://coredis.readthedocs.io/en/stable/release_notes.html?highlight=lpush#v3-0-0) -- lpush was among the functions whose signature changed from `(key, *values_to_push)` to  `(key, [values_to_push])`, and it was treating the string I'd passed it as an Iterable.  One change to `client.lpush(keyname, [string_to_push])` later and I was back up and running.

All that to say, it might be worthwhile to call out this failure mode prominently in the documentation and/or emit a warning if a String or Bytes object is passed when you expect an Iterable.  Yes, both technically qualify, but it seems to me that it's likely a vanishingly rare thing for that to be the intended behavior, and as it stands it's a nasty rake in the grass for unsuspecting new users to step on.

**Comment by alisaifee** [comment]
This problem has come up a few times and continues to make me question my choice to remove non optional variadic arguments.

There are (perhaps not highlighted well enough) a few things that have been done to help guide correct usage:
- The type annotations for arguments that expect some kind of iterable of string/bytes are explicitly defined as [Parameters](https://coredis.readthedocs.io/en/stable/api.html#coredis.typing.Parameters) which allows using a static type checker to catch exactly the error you mentioned.
- If your code base isn't ready for integration with a static type checker - the public API for coredis is (optionally) decorated with [runtime type checking](https://coredis.readthedocs.io/en/stable/handbook/typing.html#runtime-type-checking) to help catch this category of bugs during development. For example:
 ``` shell
 $ pip install coredis beartype
 $ COREDIS_RUNTIME_CHECKS=1 python -c "
import coredis
import asyncio
asyncio.run(coredis.Redis().lpush('test', 'test'))
"""
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/ali/_dev/_git/coredis/coredis/commands/_wrappers.py", line 194, in wrapped
    async with command_cache(callable, *args, **kwargs) as response:
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 199, in __aenter__
    return await anext(self.gen)
  File "/Users/ali/_dev/_git/coredis/coredis/commands/_wrappers.py", line 90, in __call__
    yield await func(*args, **kwargs)
  File "<@beartype(coredis.commands.core.CoreCommands.lpush) at 0x105d6ca60>", line 75, in lpush
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped coroutine CoreCommands.lpush() parameter elements='test' violates type hint typing.Union[list[typing.Union[str, bytes, int, float]], collections.abc.Set[typing.Union[str, bytes, int, float]], tuple[typing.Union[str, bytes, int, float], ...], collections.abc.ValuesView[typing.Union[str, bytes, int, float]], collections.abc.Iterator[typing.Union[str, bytes, int, float]]], as 'test' not <protocol "collections.abc.Set">, list, <protocol "collections.abc.ValuesView">, <protocol "collections.abc.Iterator">, or tuple.
 ```

Having said that, in the default case, nothing errors out when you call `coredis.Redis().lpush("test", "test")` and that is indeed a problem.

As a follow up I think we could do two thing:
- Improve visibility in documentation of ways to avoid the 'rake in the grass'
- Add a runtime exception if a string/byte sequence still gets through when not acceptable

**Comment by alisaifee** [comment]
Regarding documentation I've added some details regarding the high level design of the API front and centre in the README and also update the intro in the [docs](https://coredis.readthedocs.io/en/latest/#coredis) to highlight it early on.

**Comment by coandco** [comment]
Hmm.  Honestly, I think my main problem with the COREDIS_RUNTIME_CHECKS=1 approach is that it's opt-in, rather than opt-out.  I feel like you'd be much better off having "safe mode" enabled by default and allow power/production users to turn it off for performance after they've prototyped and tested their initial application, [premature optimization](http://wiki.c2.com/?PrematureOptimization) being the root of all evil and all.

That way, new users will encounter a comprehensible error message the first time they `client.lpush("test", "test")`, while those who actually need increased performance (and who are hopefully more aware of the potential pitfalls) have a simple lever to use to get it.

**Comment by coandco** [comment]
Given that the variadic->Parameter change is your main breaking change with the official `redis` module API, it might make sense to just have a function for handling Parameters, rather than having to enable global runtime type checking to avoid accidentally writing garbage data in the event of a very easy-to-make typo.

Right now, the LPUSH function's contents consist of this:
```
        return await self.execute_command(
            CommandName.LPUSH, key, *elements, callback=IntCallback()
        )
```

Theoretically, you could just add a function to validate or transform the elements, like so:
```
        return await self.execute_command(
            CommandName.LPUSH, key, *handle_parameters(elements), callback=IntCallback()
        )
```

and then have the function be a no-op if performance mode is enabled.  That would also open the possibility to having the handle_parameters function do interpretation if you wanted to -- for example, it could optionally check the type and simply wrap the Parameters argument in an array if it was a String, Bytes, or non-Iterable.

**Comment by alisaifee** [comment]
re: `COREDIS_RUNTIME_CHECKS` being enabled by default and being an opt out for performance - I might couple it with a broader 'optimized' mode which disables a few other checks which add unnecessary overhead once you've got sufficient test coverage (for example the decorators which check for mutually exclusive or inclusive parameters in certain apis like [georadius](https://github.com/alisaifee/coredis/blob/master/coredis/commands/core.py#L1564)).

**Comment by coandco** [comment]
Makes sense.

The more I think about it, the more the failure mode of "oops we silently wrote garbage data" feels like an outright bug and not something a well-designed library should ever do.  Pushing responsibility onto the end user to set up static type checking doesn't feel like a solution, because most people messing with the library to see if it meets their needs won't have that set up.  The beartype runtime type checking is a step in the right direction, but it's already unambiguous what the desired action is when a user passes in a single string/bytes/non-iterable object -- throwing an exception is certainly preferable to writing garbage data, but why not just wrap it in an array and behave as the user expects?

**Comment by alisaifee** [comment]
It's not great that this typo is easy to make and can easily go unnoticed for many redis commands even if you were to check the return type. This issue exists for ~ 50 other commands so I'm okay with a compromise of raising a `TypeError` at the cost of validation.

Something like this:
```shell
$ python -c 'import coredis; import asyncio; asyncio.run(coredis.Redis().lpush("fubar", "fubar"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/ali/_dev/_git/dotfiles/_asdf/installs/python/3.10.4/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/ali/_dev/_git/coredis/coredis/commands/_validators.py", line 154, in wrapped
    raise TypeError(
TypeError: 'lpush' parameter elements='fubar' violates expected iterable of type Parameters[ValueT]
```

Implicitly transforming an input for anything other than serialization isn't something that I want to encourage (and in fact, was something that was removed as part of the 2.x->3.x departure from the `aredis`/`redis-py` API)

**Comment by coandco** [comment]
Fair enough.  Throwing a nice comprehensible error like that makes sense to me.  The only other piece I'd love to see is a line in the documentation calling out that specific error and saying that the in most cases the fix is to wrap the offending argument in an array -- that way when people Google the error they can easily find the solution right away, since it's not immediately obvious from the message itself.

**Comment by alisaifee** [comment]
Yeah I was considered something like `violates..... did you mean ["fubar"]?`

**Comment by alisaifee** [comment]
Perhaps mildly amusing, the tests for pipeline were in fact using `lpush` incorrectly ([reference](https://github.com/alisaifee/coredis/pull/96/commits/5409277d868677882e3e5d6c80e273c7fec571b6))

**Comment by coandco** [comment]
Hah!  But because they were all testing single-character strings, the results were indistinguishable from correct usage.

**Comment by alisaifee** [comment]
The additional validation is available in [4.5.5](https://github.com/alisaifee/coredis/releases/tag/4.5.5).
